### PR TITLE
zcash_client_backend: Add documentation for the `data_api` module.

### LINKED ISSUE #1209: zcash_client_backend: Add crate-level documentation of how accounts work.
> Congratulations on threading the needle of addressing my review comment while managing to avoid writing the more extensive documentation I wanted 😄 (please open an issue for writing crate-level documentation about how accounts work in `zcash_client_backend`).

_Originally posted by @str4d in https://github.com/zcash/librustzcash/pull/1202#discussion_r1508034685_

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #1209

### CODE REVIEW

File: zcash_client_backend/src/data_api/wallet.rs [hunk 1]
```diff
@@ -311,6 +341,9 @@ where
 /// [`sapling::OutputProver`]: sapling::prover::OutputProver
 #[allow(clippy::too_many_arguments)]
 #[allow(clippy::type_complexity)]
+#[deprecated(
```
Comment (nuttycom) [inline zcash_client_backend/src/data_api/wallet.rs:344]: @str4d should we go ahead and deprecate `spend` at this point?

File: zcash_client_backend/src/data_api.rs [hunk 2]
```diff
@@ -1,4 +1,63 @@
-//! Interfaces for wallet data persistence & low-level wallet utilities.
+//! Utilities for Zcash wallet construction
+//! =======================================
```
Comment (str4d) [inline zcash_client_backend/src/data_api.rs:2]: `rustdoc` uses Markdown:
```suggestion
//! # Utilities for Zcash wallet construction
```

File: zcash_client_backend/src/data_api.rs [hunk 3]
```diff
@@ -1,4 +1,63 @@
...
+//!
+//! This module defines a set of APIs for wallet data persistence, and provides a suite of methods
+//! based upon these APIs that can be used to implement a fully functional Zcash wallet. At
+//! present, the interfaces provided here are built primarily around the use of a source of
+//! [`CompactBlock`] data such as the Zcash Light Client Protocol as defined in 
+//! [ZIP 307](https://zips.z.cash/zip-0307) but they may be generalized to full-block use cases in
+//! the future.
+//!
+//! Important Concepts
+//! ------------------
```
Comment (str4d) [inline zcash_client_backend/src/data_api.rs:12]:
```suggestion
//! ## Important Concepts
```

File: zcash_client_backend/src/data_api.rs [hunk 4]
```diff
@@ -1,4 +1,63 @@
...
+//!   details.
+//! * Witness Updates: In order to spend a shielded note, the wallet must be able to compute the
+//!   Merkle path to that note in the global note commitment tree. When [`scan_cached_blocks`] is
+//!   used to process a range of blocks, the note commitment tree is updated with the note
+//!   commitments for the blocks in that range. 
+//! * Transaction Construction: The [`wallet`] module provides functions for creating Zcash
+//!   transactions that spend funds belonging to the wallet.
+//!
+//! Core Traits
+//! -----------
```
Comment (str4d) [inline zcash_client_backend/src/data_api.rs:34]:
```suggestion
//! ## Core Traits
```

File: zcash_client_backend/src/data_api.rs [hunk 5]
```diff
@@ -1,4 +1,63 @@
...
+//!
+//! The utility functions described above depend upon four important traits defined in this
+//! module, which between them encompass the data storage requirements of a light wallet.
+//! The relevant traits are [`InputSource`], [`WalletRead`], [`WalletWrite`], and
+//! [`WalletCommitmentTrees`]. A complete implementation of the data storage layer for a wallet
+//! will include an implementation of all four of these traits. See the [`zcash_client_sqlite`]
+//! crate for a complete example of the implementation of these traits.
+//!
+//! Accounts
+//! --------
```
Comment (str4d) [inline zcash_client_backend/src/data_api.rs:44]:
```suggestion
//! ## Accounts
```

File: zcash_client_backend/src/data_api/wallet.rs [hunk 6]
```diff
@@ -1,3 +1,33 @@
+//! Functions for creating Zcash transactions that spend funds belonging to the wallet
+//! ==================================================================================
```
Comment (str4d) [inline zcash_client_backend/src/data_api/wallet.rs:2]:
```suggestion
//! # Functions for creating Zcash transactions that spend funds belonging to the wallet
```

File: zcash_client_backend/src/data_api.rs [hunk 7]
```diff
@@ -1,4 +1,63 @@
...
+//!
+//! There are several important operations that a Zcash wallet must perform that distinguish Zcash
+//! wallet design from wallets for other cryptocurrencies. 
+//!
+//! * Viewing Keys: Wallets based upon this module are built around the capabilities of Zcash
+//!   [`UnifiedFullViewingKey`]s; the wallet backend provides no facilities for the storage 
+//!   of spending keys, and spending keys must be provided by the caller in order to perform
+//!   transaction creation operations.
+//! * Blockchain Scanning: A Zcash wallet must download and trial-decrypt each transaction on the
+//!   Zcash blockchain using one or more Viewing Keyw in order to find new shielded transaction
```
Comment (str4d) [inline zcash_client_backend/src/data_api.rs:22]:
```suggestion
//!   Zcash blockchain using one or more Viewing Keys in order to find new shielded transaction
```

File: zcash_client_backend/src/data_api/wallet.rs [hunk 8]
```diff
@@ -1,3 +1,33 @@
...
+//!
+//! This module contains several different ways of creating Zcash transactions. This module is
+//! designed around the idea that a Zcash wallet holds its funds in notes in either the `orchard`
+//! or `sapling` shielded pool. In order to better preserve users' privacy, it does not provide any
```
Comment (str4d) [inline zcash_client_backend/src/data_api/wallet.rs:6]:
```suggestion
//! designed around the idea that a Zcash wallet holds its funds in notes in either the Orchard
//! or Sapling shielded pool. In order to better preserve users' privacy, it does not provide any
```

File: zcash_client_backend/src/data_api/wallet.rs [hunk 9]
```diff
@@ -1,3 +1,33 @@
...
+//! functionality that allows users to directly spend transparent funds except by sending them to a
+//! shielded internal address belonging to their wallet.
+//!
+//! The important high-level operations provided by this module are [`propose_transfer`],
+//! [`propose_shielding`], and [`create_proposed_transactions`]. 
+//!
+//! [`propose_transfer`] takes a [`TransactionRequest`] object and selects inputs notes and
+//! computes the fees required to satisfy that request, and returns a [`Proposal`] object that
+//! describes the transaction to be made.
```
Comment (str4d) [inline zcash_client_backend/src/data_api/wallet.rs:15]:
```suggestion
//! [`propose_transfer`] takes a [`TransactionRequest`] object, selects inputs notes and
//! computes the fees required to satisfy that request, and returns a [`Proposal`] object that
//! describes the transaction to be made.
```

File: zcash_client_backend/src/data_api/wallet.rs [hunk 10]
```diff
@@ -1,3 +1,33 @@
...
+//! [`create_proposed_transactions`] constructs one or more Zcash [`Transaction`]s based upon a
+//! provided [`Proposal`], stores them to the wallet database, and returns the [`TxId`] for each
+//! constructed transaction to the caller. The caller can then use the
+//! [`WalletRead::get_transaction`] method to retrieve the newly constructed transactions. It is
+//! the responsibility of the caller to retrieve and serialize the transactions and submit them for
+//! inclusion into the Zcash blockchain.
+//!
+//! [`TransactionRequest`]: crate::zip321::TransactionRequest
+//! [`propose_transfer`]: crate::data_api::wallet::propose_transfer
+//! [`propose_shielding`]: crate::data_api::wallet::propose_shielding
```
Comment (str4d) [inline zcash_client_backend/src/data_api/wallet.rs:30]:
```suggestion
//! [`propose_shielding`]: crate::data_api::wallet::propose_shielding

```

File: zcash_client_backend/src/data_api/scanning.rs [hunk 11]
```diff
@@ -1,3 +1,4 @@
+//! Common types used for managing a queue of scanning ranges.
```
Comment (str4d) [inline zcash_client_backend/src/data_api/scanning.rs:1]:
```suggestion
//! Common types used for managing a queue of scanning ranges.

```

**Review by str4d — CHANGES_REQUESTED** [review]

File: zcash_client_backend/src/data_api/wallet.rs (hunk 1 above)
Comment (str4d) [reply zcash_client_backend/src/data_api/wallet.rs:344]: Yeah, let's do that in this release. Add this to the changelog.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/bb4f00b819c0bf3980b9d55a1e19fa07fe07ffc3..e44ddee8ad27ecc68557cd7426cb5ef951b44598) to address review comments.

**Review by str4d — APPROVED** [review]
ACK e44ddee8ad27ecc68557cd7426cb5ef951b44598

**Review by str4d — APPROVED** [review]
utACK 9ea027150a38645616d942103d2b77f2a935411e

**Pull request merged**